server/src/main/java/org/elasticsearch/health/node/ShardLimitsHealthIndicatorService.java

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.health.node;
+
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.health.Diagnosis;
+import org.elasticsearch.health.HealthIndicatorDetails;
+import org.elasticsearch.health.HealthIndicatorImpact;
+import org.elasticsearch.health.HealthIndicatorResult;
+import org.elasticsearch.health.HealthIndicatorService;
+import org.elasticsearch.health.HealthStatus;
+import org.elasticsearch.health.ImpactArea;
+import org.elasticsearch.health.metadata.HealthMetadata;
+import org.elasticsearch.indices.ShardLimitValidator;
+
+import java.util.List;
+import java.util.stream.Stream;
+
+import static org.elasticsearch.indices.ShardLimitValidator.FROZEN_GROUP;
+import static org.elasticsearch.indices.ShardLimitValidator.NORMAL_GROUP;
+
+/**
+ *  This indicator reports health data about the shard limit across the cluster.
+ *
+ * <p>
+ * The indicator will report:
+ * * RED when there's room for less than 5 shards (either normal or frozen nodes)
+ * * YELLOW when there's room for less than 10 shards (either normal or frozen nodes)
+ * * GREEN otherwise
+ * </p>
+ *
+ *  Although the `max_shard_per_node(.frozen)?` information is scoped by Node, we use the information from master because there is where
+ *  the available room for new shards is checked before creating new indices.
+ */
+public class ShardLimitsHealthIndicatorService implements HealthIndicatorService {
+
+    private static final String NAME = "shard_limits";
+    private final String UPGRADE_BLOCKED = "The cluster has too many used shards to be able to upgrade";
```

Comment:
If 10 shards (YELLOW status) are still available - is the upgrade blocked already? 
I guess another way to ask this - is this impact worded correctly for both YELLOW and RED statuses? Or do we need multiple impacts?

Reply:
mmmh you're right. I did assume that YELLOW is as that bad as RED. Maybe Yellow should be more "hey, __there could be_ issues upgrading in this case", so yes, we'd need 2 impacts.